src/Silicon.cpp

```diff
@@ -825,89 +954,311 @@ namespace galsim {
+        // many arguments for the OpenMP kernel on GPU (at least with the Clang
+        // runtime)
+        std::vector<double> randomValues(nphotons * 4);
 
         UniformDeviate ud(rng);
         GaussianDeviate gd(ud, 0, 1);
 
         for (int i=0; i<nphotons; i++) {
-            diffStepRandom[i*2] = gd();
-            diffStepRandom[i*2+1] = gd();
-            pixelNotFoundRandom[i] = ud();
-            conversionDepthRandom[i] = ud();
+            randomValues[i*4] = gd();    // diffStep x
+            randomValues[i*4+1] = gd();  // diffstep y
+            randomValues[i*4+2] = ud();  // pixel not found
+            randomValues[i*4+3] = ud();  // conversion depth
         }
 
         const double invPixelSize = 1./_pixelSize; // pixels/micron
         const double diffStep_pixel_z = _diffStep / (_sensorThickness * _pixelSize);
         Bounds<int> b = target.getBounds();
         double addedFlux = 0.;
 
+        // Get everything out of C++ classes and into arrays/structures suitable for GPU
+        // photons
+        const double* photonsX = photons.getXArray();
+        const double* photonsY = photons.getYArray();
+        const double* photonsDXDZ = photons.getDXDZArray();
+        const double* photonsDYDZ = photons.getDYDZArray();
+        const double* photonsFlux = photons.getFluxArray();
+        const double* photonsWavelength = photons.getWavelengthArray();
+        bool photonsHasAllocatedAngles = photons.hasAllocatedAngles();
+        bool photonsHasAllocatedWavelengths = photons.hasAllocatedWavelengths();
+
+        // if no wavelengths allocated, photonsWavelength will be null, but some
+        // compilers don't like mapping a null pointer to the GPU, so assign it to
+        // something safe instead. (It will never be accessed in this case so the
+        // content of what it points to doesn't matter).
+        if (!photonsHasAllocatedWavelengths) {
+            photonsWavelength = photonsX;
+        }
+        // same with the angles
+        if (!photonsHasAllocatedAngles) {
+            photonsDXDZ = photonsX;
+            photonsDYDZ = photonsY;
+        }
+        
+        // random array
+        double* randomArray = randomValues.data();
+        
+        // delta image
+        int deltaXMin = _delta.getXMin();
+        int deltaYMin = _delta.getYMin();
+        int deltaXMax = _delta.getXMax();
+        int deltaYMax = _delta.getYMax();
+        int deltaStep = _delta.getStep();
+        int deltaStride = _delta.getStride();
+
+        int emptypolySize = _emptypoly.size();
+
+        double* deltaData = _delta.getData();
+        Bounds<double>* pixelInnerBoundsData = _pixelInnerBounds.data();
+        Bounds<double>* pixelOuterBoundsData = _pixelOuterBounds.data();
+        Position<float>* horizontalBoundaryPointsData = _horizontalBoundaryPoints.data();
+        Position<float>* verticalBoundaryPointsData = _verticalBoundaryPoints.data();
```

Comment:
That's correct - I tested accessing the vectors directly on GPU but it fails with an invalid memory access. I think this is because mapping the vector to the GPU only maps the vector structure and not the underlying data array (which we can't map because it's private to the vector object). So I've added a comment to explain this.